compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,341 @@
+        (testClass.name ++ "$scalajs$junit$bootstrapper").toTermName,
+        Synthetic, Synthetic,
+        List(defn.ObjectType, mydefn.BootstrapperType), newScope,
+        coord = testClass.span, assocFile = testClass.assocFile).entered
+    val classSym = moduleSym.moduleClass.asClass
+
+    val constr = genConstructor(classSym)
+
+    val testMethods = annotatedMethods(testClass, mydefn.TestAnnotClass)
+
+    val defs = List(
+        genCallOnModule(classSym, Names.beforeClass, testClass.companionModule, mydefn.BeforeClassAnnotClass),
+        genCallOnModule(classSym, Names.afterClass, testClass.companionModule, mydefn.AfterClassAnnotClass),
+        genCallOnParam(classSym, Names.before, testClass, mydefn.BeforeAnnotClass),
+        genCallOnParam(classSym, Names.after, testClass, mydefn.AfterAnnotClass),
+        genTests(classSym, testMethods),
+        genInvokeTest(classSym, testClass, testMethods),
+        genNewInstance(classSym, testClass)
+    )
+
+    if (ctx.sbtCallback != null)
+      sbt.APIUtils.registerDummyClass(classSym)
+
+    ClassDef(classSym, constr, defs)
+  }
+
+  private def genConstructor(owner: ClassSymbol)(implicit ctx: Context): DefDef = {
+    val sym = ctx.newDefaultConstructor(owner).entered
+    DefDef(sym, {
+      Block(
+        Super(This(owner), nme.EMPTY.toTypeName, inConstrCall = true).select(defn.ObjectClass.primaryConstructor).appliedToNone :: Nil,
+        unitLiteral
+      )
+    })
+  }
+
+  private def genCallOnModule(owner: ClassSymbol, name: TermName, module: Symbol, annot: Symbol)(implicit ctx: Context): DefDef = {
+    val sym = ctx.newSymbol(owner, name, Synthetic | Method,
+        MethodType(Nil, Nil, defn.UnitType)).entered
+
+    DefDef(sym, {
+      if (module.exists) {
+        val calls = annotatedMethods(module.moduleClass.asClass, annot)
+          .map(m => Apply(ref(module).select(m), Nil))
+        Block(calls, unitLiteral)
+      } else {
+        unitLiteral
+      }
+    })
+  }
+
+  private def genCallOnParam(owner: ClassSymbol, name: TermName, testClass: ClassSymbol, annot: Symbol)(implicit ctx: Context): DefDef = {
+    val sym = ctx.newSymbol(owner, name, Synthetic | Method,
+        MethodType(Names.instance :: Nil, defn.ObjectType :: Nil, defn.UnitType)).entered
+
+    DefDef(sym, { (paramRefss: List[List[Tree]]) =>
+      val List(List(instanceParamRef)) = paramRefss
+      val calls = annotatedMethods(testClass, annot)
+        .map(m => Apply(instanceParamRef.cast(testClass.typeRef).select(m), Nil))
+      Block(calls, unitLiteral)
+    })
+  }
+
+  private def genTests(owner: ClassSymbol, tests: List[Symbol])(
+      implicit ctx: Context, mydefn: MyDefinitions): DefDef = {
+
+    val sym = ctx.newSymbol(owner, Names.tests, Synthetic | Method,
+        MethodType(Nil, defn.ArrayOf(mydefn.TestMetadataType))).entered
+
+    DefDef(sym, {
+      val metadata = for (test <- tests) yield {
+        val name = Literal(Constant(test.name.toString))
```

Comment:
Is this correct with respect to name-mangling ? More tests please :).

Reply:
More tests will come from upstream Scala.js, notably coming from https://github.com/scala-js/scala-js/tree/v1.0.0-M7/test-suite/js/src/test/scala/org/scalajs/testsuite/junit, once #6365 is merged and I get to enable those tests. See for example `JUnitNamesTest` for the name-mangling aspect.